File: extraction/usda/R/psd_api_usage.r
# Required Libraries
library(httr)    # For HTTP requests
library(jsonlite) # For JSON data manipulation
library(dplyr)   # For data manipulation
library(purrr)   # For functional programming tools
library(readr)   # For writing CSV files
library(stringr) # For string manipulation

# Constants
API_KEY <- '697486e5-932d-46d3-804a-388452a19d70'
BASE_URL <- 'https://apps.fas.usda.gov/OpenData/api/psd/'
OUTPUT_FILE <- 'data/usda/commodity_forecast_2024.csv'

# CSV Headers
CABECALHOS <- c(
  "commodityCode", "countryCode", "marketYear", "calendarYear", "month",
  "attributeId", "unitId", "commodityName", "attributeName", "unitDescription",
  "countryName", "regionName", "value"
)

# Commodity Codes and Year
COMMODITIES <- c("0440000", "2631000", "0813700", "0813300", "0814200", 
                 "0813800", "0813200", "0813600", "0813100", "0813500", 
                 "4242000", "4233000", "4235000", "4243000", "4244000", 
                 "4234000", "4239100", "4232000", "4236000", "2231000", 
                 "2223000", "2232000", "2221000", "2226000", "2222000", 
                 "2224000", "0410000")
ANO <- 2024

# Function to perform GET requests
get_api_data <- function(endpoint) {
  url <- paste0(BASE_URL, endpoint)
  response <- GET(url, add_headers('Accept' = 'application/json', 'API_KEY' = API_KEY))
  stop_for_status(response)  # Check if the request was successful
  content(response, as = "text") %>% fromJSON(flatten = TRUE)  # Return the JSON content as a list
}

# Function to retrieve auxiliary data
fetch_auxiliary_data <- function() {
  regioes <- get_api_data('regions') %>% 
    with(setNames(regionName, regionCode))
  
  paises <- get_api_data('countries') %>% 
    mutate(regionName = regioes[regionCode]) %>% 
    select(countryCode, countryName, regionName)
  
  commodities <- get_api_data('commodities') %>% 
    with(setNames(commodityName, commodityCode))
  
  atributos <- get_api_data('commodityAttributes') %>% 
    mutate(attributeId = as.character(attributeId),
           attributeName = str_trim(attributeName)) %>%
    with(setNames(attributeName, attributeId))
  
  unidades <- get_api_data('unitsOfMeasure') %>% 
    mutate(unitId = as.character(unitId)) %>%
    with(setNames(str_trim(unitDescription), unitId))
  
  # Debugging: Print out the auxiliary data to check for completeness and correctness
  message("Auxiliary data fetched: ")
  print(atributos)
  
  list(paises = paises, commodities = commodities, atributos = atributos, unidades = unidades)
}

# Function to fetch forecast data
fetch_forecast_data <- function(commodity_code, ano) {
  dados_paises <- get_api_data(paste0('commodity/', commodity_code, '/country/all/year/', ano))
  dados_mundiais <- get_api_data(paste0('commodity/', commodity_code, '/world/year/', ano))
  
  # Combine country-level and world-level data
  bind_rows(dados_paises, dados_mundiais)
}

# Function to process and save the data
process_and_write_data <- function(paises, commodities, atributos, unidades) {
  write_csv(tibble(), OUTPUT_FILE, col_names = FALSE)  # Create an empty file
  write_lines(paste(CABECALHOS, collapse = ","), OUTPUT_FILE)  # Write headers
  
  walk(COMMODITIES, function(commodity_code) {
    forecast_data <- fetch_forecast_data(commodity_code, ANO) %>%
      mutate(attributeId = as.character(attributeId))  # Ensure IDs are treated as characters
    
    # Debugging: Print unmatched IDs before the join
    unmatched_ids <- setdiff(unique(forecast_data$attributeId), names(atributos))
    if (length(unmatched_ids) > 0) {
      message("Unmatched attribute IDs found: ", paste(unmatched_ids, collapse = ", "))
    }
    
    # Join forecast data with auxiliary data
    forecast_data <- forecast_data %>%
      left_join(paises, by = "countryCode") %>%
      mutate(
        commodityName = commodities[commodityCode],
        attributeName = atributos[attributeId],
        unitDescription = unidades[unitId],
        countryName = if_else(countryCode == "00", "World", countryName),
        regionName = if_else(countryCode == "00", "Global", regionName)
      )
    
    # Debugging: Check for NA values in attributeName after the join
    na_attributes <- forecast_data %>% filter(is.na(attributeName))
    if (nrow(na_attributes) > 0) {
      message("NA attributeNames found after join:")
      print(na_attributes)
    }
    
    # Select columns and save to CSV
    forecast_data %>%
      select(all_of(CABECALHOS)) %>%
      write_csv(OUTPUT_FILE, append = TRUE, col_names = FALSE)
  })
}

# Main function
main <- function() {
  dados_auxiliares <- fetch_auxiliary_data()
  process_and_write_data(dados_auxiliares$paises, dados_auxiliares$commodities, 
                         dados_auxiliares$atributos, dados_auxiliares$unidades)
  message(sprintf("Data successfully saved to %s", OUTPUT_FILE))
}

# Entry point
if (interactive()) {
  main()
}

File: extraction/usda/Python/psd_api_usage.py
import requests
import csv

# Constants
API_KEY = '697486e5-932d-46d3-804a-388452a19d70'
BASE_URL = 'https://apps.fas.usda.gov/OpenData/api/psd/'
OUTPUT_FILE = 'data/usda/commodity_forecast_2024.csv'

# Headers for the CSV output
HEADERS = [
    "commodityCode", "countryCode", "marketYear", "calendarYear", "month",
    "attributeId", "unitId", "commodityName", "attributeName", "unitDescription",
    "countryName", "regionName", "value"
]

# List of commodity codes and the year to query
COMMODITIES = ["0440000", "2631000"]
YEAR = 2024

# Helper function to make GET requests to the API
def get_api_data(endpoint):
    url = BASE_URL + endpoint
    headers = {
        'Accept': 'application/json',
        'API_KEY': API_KEY
    }
    response = requests.get(url, headers=headers)
    response.raise_for_status()
    return response.json()

# Fetch auxiliary data for regions, countries, commodities, attributes, and units
def fetch_auxiliary_data():
    regions = {item['regionCode']: item['regionName'] for item in get_api_data('regions')}
    countries = {item['countryCode']: (item['countryName'], regions.get(item['regionCode'], '')) for item in get_api_data('countries')}
    commodities = {item['commodityCode']: item['commodityName'] for item in get_api_data('commodities')}
    attributes = {item['attributeId']: item['attributeName'] for item in get_api_data('commodityAttributes')}
    units = {item['unitId']: item['unitDescription'].strip() for item in get_api_data('unitsOfMeasure')}
    
    return countries, commodities, attributes, units

# Fetch forecast data for a given commodity and year (both country and world level)
def fetch_forecast_data(commodity_code, year):
    country_data = get_api_data(f'commodity/{commodity_code}/country/all/year/{year}')
    world_data = get_api_data(f'commodity/{commodity_code}/world/year/{year}')
    
    # Merge country and world data
    all_data = country_data + world_data
    
    return all_data

# Process and write data to CSV
def process_and_write_data(countries, commodities, attributes, units):
    with open(OUTPUT_FILE, mode='w', newline='', encoding='utf-8') as file:
        writer = csv.writer(file)
        writer.writerow(HEADERS)  # Write header row
        
        for commodity_code in COMMODITIES:
            forecast_data = fetch_forecast_data(commodity_code, YEAR)
            
            for record in forecast_data:
                commodity_name = commodities.get(record['commodityCode'], '')
                attribute_name = attributes.get(record['attributeId'], '')
                unit_description = units.get(record['unitId'], '')
                
                if record['countryCode'] == '00':  # World data
                    country_name = 'World'
                    region_name = 'Global'
                else:  # Country data
                    country_name, region_name = countries.get(record['countryCode'], ('', ''))
                
                row = [
                    record['commodityCode'], record['countryCode'], record['marketYear'], record['calendarYear'],
                    record['month'], record['attributeId'], record['unitId'], commodity_name, attribute_name,
                    unit_description, country_name, region_name, record['value']
                ]
                
                writer.writerow(row)

# Main function to run the script
def main():
    # Fetch auxiliary data
    countries, commodities, attributes, units = fetch_auxiliary_data()
    
    # Process and write data to CSV
    process_and_write_data(countries, commodities, attributes, units)
    
    print(f"Data successfully written to {OUTPUT_FILE}")

# Entry point of the script
if __name__ == "__main__":
    main()
